File: scoreWLDstat.py
import concurrent.futures
from collections import Counter
import chess
import chess.pgn
import re
import json
import os
import argparse
import io
from multiprocessing import cpu_count
import time
import gc


def chunks(lst, n):
    """Yield successive n-sized chunks from lst."""
    for i in range(0, len(lst), n):
        yield lst[i : i + n]


class PosAnalyser:
    def __init__(self, plies):
        self.matching_plies = plies

    def ana_pos(self, pgn, offsets):
        matstats = Counter()

        p = re.compile("([+-]*M*[0-9.]*)/([0-9]*)")
        mateRe = re.compile("([+-])M[0-9]*")

        for offset in offsets:
            # read game
            pgn.seek(offset)
            game = chess.pgn.read_game(pgn)
            if game == None:
                break

            # get result
            result = game.headers["Result"]
            if result == "1/2-1/2":
                resultkey = "D"
            elif result == "1-0":
                resultkey = "W"
            elif result == "0-1":
                resultkey = "L"
            else:
                continue

            # look at the game,
            plies = 0
            board = game.board()
            for node in game.mainline():

                plies = plies + 1
                if plies > 400:
                    break
                plieskey = (plies + 1) // 2

                turn = board.turn
                scorekey = None
                m = p.search(node.comment)
                if m:
                    score = m.group(1)
                    m = mateRe.search(score)
                    if m:
                        if m.group(1) == "+":
                            score = 1001
                        else:
                            score = -1001
                    else:
                        score = int(float(score) * 100)

                        if score > 1000:
                            score = 1000
                        elif score < -1000:
                            score = -1000

                        score = (score // 5) * 5  # reduce precision

                    if turn == chess.BLACK:
                        score = -score

                    scorekey = score

                knights = bin(board.knights).count("1")
                bishops = bin(board.bishops).count("1")
                rooks = bin(board.rooks).count("1")
                queens = bin(board.queens).count("1")
                pawns = bin(board.pawns).count("1")

                matcountkey = 9 * queens + 5 * rooks + 3 * knights + 3 * bishops + pawns

                if scorekey is not None:
                    matstats[(resultkey, plieskey, matcountkey, scorekey)] += 1

                board = node.board()

        return matstats


if __name__ == "__main__":
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "--matching_plies",
        type=int,
        default=6,
        help="Number of plies that the material situation needs to be on the board (unused).",
    )

    parser.add_argument(
        "--dir", type=str, default="pgns", help="Directory with the pgns."
    )

    args = parser.parse_args()

    pgns = [args.dir + "/" + f for f in os.listdir(args.dir) if f.endswith("pgn")]

    # Combine all pgns
    combined = ""

    for file in pgns:
        with open(file) as fp:
            data = fp.read()
            combined += data
            combined += "\n"

    pgn = io.StringIO(combined)

    del combined
    gc.collect()

    offsets = []

    while True:
        # save offsets
        offset = pgn.tell()
        headers = chess.pgn.read_headers(pgn)

        if headers is None:
            break

        offsets.append(offset)

    # map sharp_pos to all pgn files using an executor
    ana = PosAnalyser(args.matching_plies)

    # split up pgns across workers
    workers = cpu_count()
    fw_ratio = len(offsets) / (4 * workers)
    offsets = list(chunks(offsets, max(1, int(fw_ratio))))

    t0 = time.time()

    futures = []
    res = Counter()
    with concurrent.futures.ProcessPoolExecutor() as e:
        for entry in offsets:
            futures.append(e.submit(ana.ana_pos, pgn, entry))
    for future in futures:
        res.update(future.result())

    t1 = time.time()

    print("Completed in:", str(t1 - t0) + " seconds")

    # and print all the fens
    with open("scoreWLDstat.json", "w") as outfile:
        json.dump(
            {
                str(k): v
                for k, v in sorted(res.items(), key=lambda item: item[1], reverse=True)
            },
            outfile,
            indent=1,
        )

File: scoreWLDstat.cpp
#include <atomic>
#include <chrono>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <functional>
#include <future>
#include <iostream>
#include <mutex>
#include <queue>
#include <stdexcept>
#include <string>
#include <thread>
#include <type_traits>
#include <unordered_map>
#include <vector>

#include "chess.hpp"
#include "json.hpp"
#include "threadpool.hpp"

using namespace chess;

namespace fs = std::filesystem;

enum class Result { WIN, LOSS, DRAW };

struct ResultKey {
    Result white;
    Result black;
};

typedef std::tuple<Result, int, int, int> map_key_t;

struct key_hash : public std::unary_function<map_key_t, std::size_t> {
    std::size_t operator()(const map_key_t &k) const {
        std::uint32_t hash = static_cast<int>(std::get<0>(k));
        hash ^= std::get<1>(k) + 0x9e3779b9 + (hash << 6) + (hash >> 2);
        hash ^= std::get<2>(k) + 0x9e3779b9 + (hash << 6) + (hash >> 2);
        hash ^= std::get<3>(k) + 0x9e3779b9 + (hash << 6) + (hash >> 2);
        return hash;
    }
};

struct key_equal : public std::binary_function<map_key_t, map_key_t, bool> {
    bool operator()(const map_key_t &v0, const map_key_t &v1) const {
        return (std::get<0>(v0) == std::get<0>(v1) && std::get<1>(v0) == std::get<1>(v1) &&
                std::get<2>(v0) == std::get<2>(v1)) &&
               std::get<3>(v0) == std::get<3>(v1);
    }
};

using map_t = std::unordered_map<map_key_t, int, key_hash, key_equal>;

class PosAnalyzer {
   public:
    [[nodiscard]] map_t pos_map(std::vector<std::string> files) {
        map_t pos_map;
        pos_map.reserve(1200000);

        for (auto file : files) {
            std::ifstream pgnFile(file);
            std::string line;

            while (true) {
                auto game = pgn::readGame(pgnFile);

                if (!game.has_value()) {
                    break;
                }

                if (game.value().headers().find("Result") == game.value().headers().end()) {
                    break;
                }

                const auto result = game.value().headers().at("Result");

                ResultKey key;

                if (result == "1-0") {
                    key.white = Result::WIN;
                    key.black = Result::LOSS;
                } else if (result == "0-1") {
                    key.white = Result::LOSS;
                    key.black = Result::WIN;
                } else if (result == "1/2-1/2") {
                    key.white = Result::DRAW;
                    key.black = Result::DRAW;
                } else {
                    continue;
                }

                int plies = 0;

                Board board = Board();

                if (game.value().headers().find("FEN") != game.value().headers().end()) {
                    board.setFen(game.value().headers().at("FEN"));
                }

                if (game.value().headers().find("Variant") != game.value().headers().end()) {
                    if (game.value().headers().at("Variant") == "fischerandom") {
                        board.set960(true);
                    }
                }

                for (const auto &move : game.value().moves()) {
                    plies++;

                    if (plies > 400) {
                        break;
                    }

                    const auto match_score = utils::splitString(move.comment, '/');

                    const int plieskey = (plies + 1) / 2;

                    int score_key = 0;

                    bool found_score = false;

                    if (match_score.size() >= 1 && move.comment != "book") {
                        found_score = true;

                        if (match_score[0][1] == 'M') {
                            if (match_score[0][1] == '+') {
                                score_key = 1001;
                            } else {
                                score_key = -1001;
                            }

                        } else {
                            const auto score = std::stof(match_score[0]);

                            int score_adjusted = score * 100;

                            if (score_adjusted > 1000) {
                                score_adjusted = 1000;
                            } else if (score_adjusted < -1000) {
                                score_adjusted = -1000;
                            }

                            score_key = int(std::floor(score_adjusted / 5.0)) * 5;
                        }
                    }

                    const auto knights = builtin::popcount(board.pieces(PieceType::KNIGHT));
                    const auto bishops = builtin::popcount(board.pieces(PieceType::BISHOP));
                    const auto rooks = builtin::popcount(board.pieces(PieceType::ROOK));
                    const auto queens = builtin::popcount(board.pieces(PieceType::QUEEN));
                    const auto pawns = builtin::popcount(board.pieces(PieceType::PAWN));

                    const int matcountkey =
                        9 * queens + 5 * rooks + 3 * bishops + 3 * knights + pawns;

                    if (found_score) {
                        const auto turn =
                            board.sideToMove() == Color::WHITE ? key.white : key.black;

                        const auto key = std::make_tuple(turn, plieskey, matcountkey, score_key);
                        pos_map[key] += 1;
                    }

                    board.makeMove(move.move);
                }
            }

            pgnFile.close();
        }

        return pos_map;
    }
};

/// @brief Get all files from a directory.
/// @param path
/// @return
[[nodiscard]] std::vector<std::string> getFiles(std::string_view path = "./pgns") {
    std::vector<std::string> files;

    for (const auto &entry : fs::directory_iterator(path)) {
        files.push_back(entry.path().string());
    }

    return files;
}

/// @brief Split into successive n-sized chunks from pgns.
/// @param pgns
/// @param targetchunks
/// @return
[[nodiscard]] std::vector<std::vector<std::string>> chunkPgns(const std::vector<std::string> &pgns,
                                                              int targetchunks) {
    std::vector<std::vector<std::string>> pgnschunked;

    int chunks_size = (pgns.size() + targetchunks - 1) / targetchunks;

    auto begin = pgns.begin();
    auto end = pgns.end();

    while (begin != end) {
        auto next =
            std::next(begin, std::min(chunks_size, static_cast<int>(std::distance(begin, end))));
        pgnschunked.push_back(std::vector<std::string>(begin, next));
        begin = next;
    }

    return pgnschunked;
}

/// @brief
/// @param argc
/// @param argv Possible ones are --dir and --file
/// @return
int main(int argc, char const *argv[]) {
    const std::vector<std::string> args(argv + 1, argv + argc);

    std::vector<std::string> files_pgn;

    if (std::find(args.begin(), args.end(), "--dir") != args.end()) {
        const auto path = std::find(args.begin(), args.end(), "--dir") + 1;
        files_pgn = getFiles(*path);
    } else if (std::find(args.begin(), args.end(), "--file") != args.end()) {
        const auto path = std::find(args.begin(), args.end(), "--file") + 1;
        files_pgn = {*path};
    } else {
        files_pgn = getFiles();
    }

    int targetchunks = 4 * std::max(1, int(std::thread::hardware_concurrency()));

    std::vector<std::vector<std::string>> files_chunked = chunkPgns(files_pgn, targetchunks);

    std::cout << "Found " << files_pgn.size() << " pgn files, creating " << files_chunked.size()
              << " chunks for processing." << std::endl;

    map_t pos_map;
    pos_map.reserve(1200000);

    // Create a thread pool
    ThreadPool pool(std::thread::hardware_concurrency());

    // Futures hold the results of each thread
    std::mutex g_i_mutex;

    const auto t0 = std::chrono::high_resolution_clock::now();

    for (const auto &files : files_chunked) {
        pool.enqueue([&files, &g_i_mutex, &pos_map]() {
            PosAnalyzer analyzer = PosAnalyzer();
            const auto map = analyzer.pos_map(files);

            {
                const std::lock_guard<std::mutex> lock(g_i_mutex);

                for (const auto &pair : map) {
                    pos_map[pair.first] += pair.second;
                }

                std::cout << "map size " << map.size() << std::endl;
            }
        });
    }

    // Wait for all threads to finish
    pool.wait();

    const auto t1 = std::chrono::high_resolution_clock::now();

    std::cout << "Time taken: " << std::chrono::duration_cast<std::chrono::seconds>(t1 - t0).count()
              << "s" << std::endl;

    // save json
    std::uint64_t total = 0;

    std::ofstream outFile("scoreWLDstat.json");

    nlohmann::json j;

    // for (const auto &pair : pos_map) {
    //     const auto map_key = "('" + pair.first[0] + "', " + std::to_string(plieskey) +
    //                          ",
    //                          " +
    //                          std::to_string(matcountkey) +
    //                          ", " + std::to_string(score_key) + ")";
    //     j[pair.first] = pair.second;
    //     total += pair.second;
    // }

    outFile << j.dump(4);

    outFile.close();

    std::cout << "Retained " << total << " scored positions for analysis." << std::endl;

    return 0;
}
